### PR TITLE
SONARJAVA-1030 First implementation inspired from close resources

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+      for (Entry<Symbol, LockedOccurence> entry : executionState.lockedOccurencesBySymbol.entrySet()) {
+        Symbol symbol = entry.getKey();
+        LockedOccurence currentOccurence = getLockedOccurence(symbol);
+        LockedOccurence occurenceToMerge = entry.getValue();
+        if (currentOccurence != null) {
+          currentOccurence.state = currentOccurence.state.merge(occurenceToMerge.state);
+          lockedOccurencesBySymbol.put(symbol, currentOccurence);
+        } else {
+          // possible way to solve the problem of variable defined in outer Execution state. lockedOccurencesBySymbol.put(symbol,
+          // occurenceToMerge);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+      this.parent = parent;
+      this.check = parent.check;
+    }
+
+    @Override
+    public String toString() {
+      String s = "";
+      for (Entry<Symbol, LockedOccurence> symbolLockedOccurenceEntry : lockedOccurencesBySymbol.entrySet()) {
+        LockedOccurence value = symbolLockedOccurenceEntry.getValue();
+        s += "  " + symbolLockedOccurenceEntry.getKey().name() + "  " + ((JavaTree) value.lastAssignment).getLine() + "  " + value.state.name() + "  ";
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java]: :new: MAJORUse a StringBuilder instead.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1643)

File: java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+        } else {
+          lockedOccurencesBySymbol.put(symbol, occurence);
+        }
+      }
+      return this;
+    }
+
+    public ExecutionState restoreParent() {
+      if (parent != null) {
+        // insertIssues();
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 3 major

Details:
* [Use a StringBuilder instead.](https://github.com/SonarSource/sonar-java/blob/353770c61e2ec3df7e310a5bd17720dc00055b70/java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java#L407)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1643)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonar-java/blob/353770c61e2ec3df7e310a5bd17720dc00055b70/java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java#L425)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonar-java/blob/353770c61e2ec3df7e310a5bd17720dc00055b70/java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java#L449)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+      }
+    }
+    return results;
+  }
+
+  @CheckForNull
+  private CloseableOccurence getCloseableOccurence(Symbol symbol) {
+    CloseableOccurence occurence = closeableOccurenceBySymbol.get(symbol);
+    if (occurence != null) {
+      return new CloseableOccurence(occurence.lastAssignment, occurence.state);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java]: :new: MAJORRemove this instantiation of "CloseableOccurence".[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)

File: java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+      return State.IGNORED;
+    } else if (isNull(expression)) {
+      return State.NULL;
+    } else if (expression.is(Tree.Kind.NEW_CLASS)) {
+      if (usesIgnoredCloseableAsArgument(((NewClassTree) expression).arguments())) {
+        return State.IGNORED;
+      }
+      return State.OPEN;
+    }
+    // TODO SONARJAVA-1029 : Engine currently ignore closeables which are retrieved from method calls. Handle them as OPEN.
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java]: :new: INFOComplete the task associated to this TODO comment.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    if (tree.is(Tree.Kind.VARIABLE)) {
+      type = ((VariableTree) tree).symbol().type();
+    } else {
+      type = ((IdentifierTree) tree).symbol().type();
+    }
+    check.addIssue(tree, "Close this \"" + type.name() + "\"");
+  }
+
+  void addCloseable(Symbol symbol, Tree lastAssignmentTree, @Nullable ExpressionTree assignmentExpression) {
+    CloseableOccurence newOccurence = new CloseableOccurence(lastAssignmentTree, getCloseableStateFromExpression(symbol, assignmentExpression));
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java]: :new: MAJORRemove this instantiation of "CloseableOccurence".[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)

File: java-checks/src/main/java/org/sonar/java/locks/LockState.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      if (s == NULL) {
+        return this;
+      }
+      return s;
+    }
+  },
+  LOCKED {
+    @Override
+    public State merge(State s) {
+      if (s == IGNORED) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/locks/LockState.java]: :new: MAJORChange this comparison to use the equals method.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/locks/LockState.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  NULL {
+    @Override
+    public State merge(State s) {
+      return s;
+    }
+  },
+  UNLOCKED {
+    @Override
+    public State merge(State s) {
+      if (s == NULL) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/locks/LockState.java]: :new: MAJORChange this comparison to use the equals method.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/locks/LockedVisitor.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  private Symbol extractSymbol(ExpressionTree expressionTree) {
+    if (expressionTree.is(Tree.Kind.MEMBER_SELECT)) {
+      return ((MemberSelectExpressionTree) expressionTree).identifier().symbol();
+    } else if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      return ((IdentifierTree) expressionTree).symbol();
+    }
+    return null;
+  }
+
+  public void insertIssues() {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/locks/LockedVisitor.java]: :new: MAJORAdd the "@Override" annotation above this method signature[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

File: java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  public void markValueAs(Symbol symbol, State state) {
+    Value value = getValue(symbol);
+    if (value != null) {
+      valuesBySymbol.put(symbol, new Value(value.treeNode, state));
+    }
+  }
+
+  public ExecutionState restoreParent() {
+    if (parent != null) {
+      // insertIssues();
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    for (Map.Entry<Symbol, Value> entry : executionState.valuesBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      Value currentValue = getValue(symbol);
+      Value valueToMerge = entry.getValue();
+      if (currentValue != null) {
+        currentValue.state = currentValue.state.merge(valueToMerge.state);
+        valuesBySymbol.put(symbol, currentValue);
+      } else {
+        // possible way to solve the problem of variable defined in outer Execution state. lockedOccurencesBySymbol.put(symbol,
+        // occurenceToMerge);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 new issues:
* 7 major
* 1 info

Details:
* [Remove this instantiation of "CloseableOccurence".](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java#L114)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)
* [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java#L140)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* [Remove this instantiation of "CloseableOccurence".](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java#L263)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L46)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L55)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Add the "@Override" annotation above this method signature](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/locks/LockedVisitor.java#L118)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java#L58)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonar-java/blob/610b536c76df1ea23f357f59c304bef9aa08c49c/java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java#L114)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 new issues:
* 4 major
* 1 info

Details:
* [Remove this instantiation of "CloseableOccurence".](https://github.com/SonarSource/sonar-java/blob/967e95a070e53685c531a5f8eea8d61e2b8742b6/java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java#L263)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)
* [Remove this instantiation of "CloseableOccurence".](https://github.com/SonarSource/sonar-java/blob/967e95a070e53685c531a5f8eea8d61e2b8742b6/java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java#L114)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)
* [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/967e95a070e53685c531a5f8eea8d61e2b8742b6/java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java#L140)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/967e95a070e53685c531a5f8eea8d61e2b8742b6/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L55)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/967e95a070e53685c531a5f8eea8d61e2b8742b6/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L46)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/symexecengine/DataFlowVisitor.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    // do nothing, inner methods will be visited later
+  }
+
+  @Override
+  public void visitClass(ClassTree tree) {
+    // do nothing, inner methods will be visited later
+  }
+
+  @Override
+  public void visitTryStatement(TryStatementTree tree) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/symexecengine/DataFlowVisitor.java]: Why ignoring the resources?

File: java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+  public ExecutionState merge(ExecutionState executionState) {
+    for (Map.Entry<Symbol, Value> entry : executionState.valuesBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      Value currentValue = getValue(symbol);
+      Value valueToMerge = entry.getValue();
+      if (currentValue != null) {
+        currentValue.state = currentValue.state.merge(valueToMerge.state);
+        valuesBySymbol.put(symbol, currentValue);
+      } else {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java]: Maybe use `else if { ... }` instead ?

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+      return CloseableState.IGNORED;
+    } else if (isNull(expression)) {
+      return CloseableState.NULL;
+    } else if (expression.is(Tree.Kind.NEW_CLASS)) {
+      if (usesIgnoredCloseableAsArgument(((NewClassTree) expression).arguments())) {
+        return CloseableState.IGNORED;
+      }
+      return CloseableState.OPEN;
+    }
+    // TODO SONARJAVA-1029 : Engine currently ignore closeables which are retrieved from method calls. Handle them as OPEN.
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java]: :new: INFOComplete the task associated to this TODO comment.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+    return expression == null || expression.is(Tree.Kind.NULL_LITERAL);
+  }
+
+  private  boolean shouldBeIgnored(Symbol symbol, @Nullable ExpressionTree expression) {
+    return shouldBeIgnored(symbol) || shouldBeIgnored(expression);
+  }
+
+  private boolean shouldBeIgnored(Symbol symbol) {
+    CloseableState state = (CloseableState) executionState.getStateOf(symbol);
+    return (state != null && state.isIgnored()) || symbol.isFinal()
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java]: :new: MAJORReduce the number of conditional operators (4) used in the expression (maximum allowed 3).[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+      if (s == NULL) {
+        return this;
+      }
+      return s;
+    }
+  },
+  OPEN {
+    @Override
+    public State merge(State s) {
+      if (s == IGNORED) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java]: :new: MAJORChange this comparison to use the equals method.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+  NULL {
+    @Override
+    public State merge(State s) {
+      return s;
+    }
+  },
+  CLOSED {
+    @Override
+    public State merge(State s) {
+      if (s == NULL) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java]: :new: MAJORChange this comparison to use the equals method.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  },
+  IGNORED {
+    @Override
+    public State merge(State s) {
+      return this;
+    }
+  };
+
+  public abstract State merge(State s);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java]: :new: MAJORAdd the "@Override" annotation above this method signature[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 new issues:
* 6 major
* 1 minor
* 1 info

Details:
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java#L55)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Add the "@Override" annotation above this method signature](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java#L73)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java#L46)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L148)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L133)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L46)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L55)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/d86e726e44823b074a2e00b7fb4f079e774e7500/java-checks/src/main/java/org/sonar/java/locks/LockedVisitor.java)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00113)

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+  }
+
+  @Override
+  public void visitDoWhileStatement(DoWhileStatementTree tree) {
+    visitStatement(tree.statement());
+    scan(tree.condition());
+  }
+
+  @Override
+  public void visitForStatement(ForStatementTree tree) {
```
Comment (merciesa) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java]: from an execution point of view visit should be done in that order:
initializer, condition, update, statement

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+      if (s.equals(NULL)) {
+        return this;
+      }
+      return s;
+    }
+  },
+  OPEN {
+    @Override
+    public State merge(State s) {
+      if (s.equals(IGNORED)) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java]: :new: MAJORChange this comparison to use the equals method.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+  NULL {
+    @Override
+    public State merge(State s) {
+      return s;
+    }
+  },
+  CLOSED {
+    @Override
+    public State merge(State s) {
+      if (s.equals(NULL)) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java]: :new: MAJORChange this comparison to use the equals method.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

File: java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  },
+  IGNORED {
+    @Override
+    public State merge(State s) {
+      return this;
+    }
+  };
+
+  public boolean isIgnored() {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java]: :new: MAJORAdd the "@Override" annotation above this method signature[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 new issues:
* 6 major
* 1 minor
* 1 info

Details:
* [Add the "@Override" annotation above this method signature](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java#L73)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java#L55)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/closeresource/CloseableState.java#L46)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L148)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L133)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L55)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L46)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/locks/LockedVisitor.java)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 new issues:
* 3 major
* 1 minor
* 1 info

Details:
* [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L133)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L148)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L46)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/locks/LockState.java#L55)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)
* [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/5190cb8b44121afddf5c452a4b3754bd1ea95ff8/java-checks/src/main/java/org/sonar/java/locks/LockedVisitor.java)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 1 major
* 1 info

Details:
* [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/59f1d926caabfdc5c3bce14acb003ccc73646a77/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L133)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)
* [Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).](https://github.com/SonarSource/sonar-java/blob/59f1d926caabfdc5c3bce14acb003ccc73646a77/java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java#L148)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

**Pull request closed without merging**